x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java

```diff
@@ -0,0 +1,298 @@
+        logger.debug("Policy [{}]: Validating [{}] source mappings", policyName, sourceIndices);
+        for (String sourceIndex : sourceIndices) {
+            Map<String, Object> mapping = getMappings(getIndexResponse, sourceIndex);
+            Map<?, ?> properties = ((Map<?, ?>) mapping.get("properties"));
+            if (properties == null) {
+                listener.onFailure(
+                    new ElasticsearchException(
+                        "Enrich policy execution for [{}] failed. Could not read mapping for source [{}] included by pattern [{}]",
+                        policyName, sourceIndex, policy.getIndexPattern()));
+            }
+            if (properties.containsKey(policy.getEnrichKey()) == false) {
+                listener.onFailure(
+                    new ElasticsearchException(
+                        "Enrich policy execution for [{}] failed. Could not locate enrich key field [{}] on mapping for index [{}]",
+                        policyName, policy.getEnrichKey(), sourceIndex));
+            }
+            for (String enrichField : policy.getEnrichValues()) {
+                if (properties.containsKey(enrichField) == false) {
+                    listener.onFailure(new ElasticsearchException(
+                        "Enrich policy execution for [{}] failed. Could not locate enrich value field [{}] on mapping for index [{}]",
+                        policyName, enrichField, sourceIndex));
+                }
+            }
+        }
+    }
+
+    private String getEnrichIndexBase(final String policyName) {
+        return ".enrich-" + policyName;
+    }
+
+    private Map<String, Object> resolveEnrichMapping(final EnrichPolicy policy) {
+        // TODO: Modify enrichKeyFieldMapping based on policy type.
+        // Create dynamic mapping templates - one for the enrich key and one for the enrich values
+        Map<String, Object> enrichKeyFieldMapping = MapBuilder.<String, Object>newMapBuilder()
+            .put("type", "{dynamic_type}")
+            .put("indexed", true)
+            .put("doc_values", false)
+            .map();
+
+        Map<String, Object> enrichKeyTemplateDefinition = MapBuilder.<String, Object>newMapBuilder()
+            .put("match", policy.getEnrichKey())
+            .put("mapping", enrichKeyFieldMapping)
+            .map();
+
+        Map<String, Object> enrichKeyTemplate = MapBuilder.<String, Object>newMapBuilder()
+            .put("enrich_key_template", enrichKeyTemplateDefinition)
+            .map();
+
+        Map<String, Object> enrichTextValueFieldMapping = MapBuilder.<String, Object>newMapBuilder()
+            .put("type", "keyword")
+            .put("ignore_above", 256)
+            .put("indexed", false)
+            .put("doc_values", true)
+            .map();
+
+        Map<String, Object> enrichTextValueTemplateDefinition = MapBuilder.<String, Object>newMapBuilder()
+            .put("match_mapping_type", "string")
```

Comment:
Why do we match on Strings ? (if it is for the 256 limit, we should allow larger for values and smaller for keys... possibly a todo to expose these as soft limits ?)

Reply:
This is for the case that an enrich value is mapped as a `text` field. Copying a `text` field as-is from a source index does not work since `doc_values` [cannot be enabled for them.](https://www.elastic.co/guide/en/elasticsearch/reference/current/doc-values.html)

Reply:
The 256 limit simply came from the default keyword mapping. I'm fine with modifying that.